neo/Consensus/ConsensusContext.cs

```diff
@@ -200,6 +239,28 @@ public ConsensusPayload MakePrepareRequest()
             });
         }
 
+        public ConsensusPayload MakeRecoveryMessage()
+        {
+            var changeViewPayloads = TransactionHashes == null || PreparationPayloads.Count(p => p != null) < M
+                ? ChangeViewPayloads : null;
+
+            return MakeSignedPayload(new RecoveryMessage()
+            {
+                ChangeViewWitnessInvocationScripts = changeViewPayloads?.Select(p => p.Witness.InvocationScript).ToArray(),
+                ChangeViewTimestamps = changeViewPayloads?.Select(p => p.Timestamp).ToArray(),
+                OriginalChangeViewNumbers = changeViewPayloads?.Select(p => p.GetDeserializedMessage<ChangeView>().NewViewNumber).ToArray(),
+                TransactionHashes = TransactionHashes,
+                Nonce = Nonce,
+                NextConsensus = NextConsensus,
+                MinerTransaction = (MinerTransaction) (TransactionHashes == null ? null : Transactions?[TransactionHashes[0]]),
+                // We only need a PreparationHash set if we don't have the PrepareRequest information.
+                PreparationHash = TransactionHashes == null ? PreparationPayloads.Where(p => p != null).GroupBy(p => p.Hash, (k, g) => new { Hash = k, Count = g.Count() }).OrderByDescending(p => p.Count).Select(p => p.Hash).FirstOrDefault() : null,
```

Comment:
I think in this place we should use `PrepareResponse.PreparationHash` instead of `p.Hash`.

Reply:
I'll take a deeper look. One thing to consider is that .not all of them are of type `PrepareResponse` one of them is `PrepareRequest`

Reply:
That makes sense.

Reply:
Nice catch, I fixed it.